File: objectivec/GithubappObjC/RepoSearchViewController.m
//
//  TableViewController.m
//  GithubappObjC
//
//  Created by user124076 on 05.01.17.
//  Copyright © 2017 user124076. All rights reserved.
//

#import "RepoSearchViewController.h"

@interface RepoSearchViewController () <UISearchBarDelegate>


@end

@implementation RepoSearchViewController

-(void) awakeFromNib{
    [super awakeFromNib];
    
    [RepoSearchAssembly configure:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //todo: naci negdje drugo mjesto gdje bi se moglo deklarirati
    self.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
    self.selectedSortMethod = @"stars";
    
    self.searchController.searchBar.delegate = self;
    self.searchController.dimsBackgroundDuringPresentation = false;
    self.definesPresentationContext = true;
    
    self.tableView.tableHeaderView = self.searchController.searchBar;
    
}

-(void)addRepos:(NSArray *)repos{
    self.repos = repos;
    
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
        dispatch_async(dispatch_get_main_queue(), ^(void){
            [[self tableView] reloadData];
        });
    });
}

-(void) showErrorMessage:(NSError *)error{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:[error localizedDescription] preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];

    [self presentViewController:alertController animated:true completion:nil];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    
    [self.presenter fetchRepos:searchBar.text withSortMethod:self.selectedSortMethod];
}

-(IBAction)segmentedControlValueChanged:(UISegmentedControl*)sender{
    switch (sender.selectedSegmentIndex) {
        case 0:  //stars selected
            self.selectedSortMethod = @"stars";
            
            self.repos = [self.repos sortedArrayUsingComparator:^NSComparisonResult(NSDictionary *d1, NSDictionary *d2){
                
                return (int)d1[@"stargazers_count"] < (int)d2[@"stargazers_count"];
                
            }];
            break;
        case 1:   //forks selected
            self.selectedSortMethod = @"forks";
            
            self.repos = [self.repos sortedArrayUsingComparator:^NSComparisonResult(NSDictionary *d1, NSDictionary *d2){
                
                return (int)d1[@"forks_count"] < (int)d2[@"forks_count"];
                
            }];
            break;
        default:  //date selected
            self.selectedSortMethod = @"updated";
            
            self.repos = [self.repos sortedArrayUsingComparator:^NSComparisonResult(NSDictionary *d1, NSDictionary *d2){
                
                NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
                dateFormatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ssZ";
                NSDate *date1 = [dateFormatter dateFromString:d1[@"pushed_at"]];
                NSDate *date2 = [dateFormatter dateFromString:d2[@"pushed_at"]];
                
                NSComparisonResult result = [date1 compare:date2];
                
                switch (result)
                {
                    case NSOrderedAscending: return true;
                    default: return false;
                }
                
            }];
    }
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
        dispatch_async(dispatch_get_main_queue(), ^(void){
            [[self tableView] reloadData];
        });
    });
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [[self repos]count];
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return [[UIScreen mainScreen] bounds].size.height * 0.1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    RepoCell *cell =[tableView dequeueReusableCellWithIdentifier:@"RepoCell" forIndexPath:indexPath];
    
    NSDictionary* repo = self.repos[indexPath.row];
    
    cell.repoNameLabel.text = repo[@"name"];
    cell.userNameLabel.text = repo[@"owner"][@"login"];
    
    cell.forkNoLabel.text = [NSString stringWithFormat:@"%@ forks", repo[@"forks_count"]];
    cell.issueNoLabel.text = [NSString stringWithFormat:@"%@ issues", repo[@"open_issues"]];
    cell.watcherNoLabel.text = [NSString stringWithFormat:@"%@ watchers", repo[@"watchers_count"]];
    
    [cell.userThumbImageView sd_setImageWithURL:[NSURL URLWithString:repo[@"owner"][@"avatar_url"]]
                 placeholderImage:[UIImage imageNamed:@"placeholder-user"]];
    
    cell.userThumbImageView.userInteractionEnabled = true;
    cell.userThumbImageView.tag = indexPath.row;
    UITapGestureRecognizer *tapped = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tappedOnImage:)];
    tapped.numberOfTapsRequired = 1;
    [cell.userThumbImageView addGestureRecognizer:tapped];
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [self.presenter gotoRepoDetailScreen];
}

-(void)tappedOnImage:(UITapGestureRecognizer*)sender{
    UIImageView *senderImageView = (UIImageView *)sender.view;
    self.selectedIndexPath = senderImageView.tag;
    [self.presenter gotoUserDetailScreen];
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    [[self presenter] passDataToNextScene:segue];
}


@end

File: swift/GithubApp/RepoSearchViewController.swift
//
//  ViewController.swift
//  GitHub-app
//
//  Created by user124076 on 1/2/17.
//  Copyright © 2017 user124076. All rights reserved.
//
import UIKit
import Kingfisher

protocol RepoSearchViewControllerOutput {
    func fetchRepos(query: String, sortMethod: String)
    func gotoRepoDetailScreen()
    func gotoUserDetailScreen()
    func passDataToNextScene(_ segue: UIStoryboardSegue)
}

protocol RepoSearchViewControllerInput: class{
    func displayFetchedRepos(_ repos: [Repo])
    func displayErrorView(_ errorMessage: String)
    
    func showWaitingView()
    func hideWaitingView()
}

class RepoSearchViewController: UITableViewController {
    
    let searchController = UISearchController(searchResultsController : nil)
    
    var presenter: RepoSearchViewControllerOutput!
    
    var indexPathSelected: IndexPath = IndexPath()
    var selectedSortMethod:String = "stars"
    
    var repos = [Repo]()
    
    override func awakeFromNib() {
        super.awakeFromNib()
        RepoSearchAssembly.sharedInstance.configure(self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        searchController.searchBar.delegate = self
        searchController.dimsBackgroundDuringPresentation = false
        definesPresentationContext = true
        tableView.tableHeaderView = searchController.searchBar

    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return repos.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: RepoCell.defaultReuseIdentifier, for: indexPath) as! RepoCell
        
        let repo = repos[indexPath.row]
        cell.repoNameLabel.text = repo.name
        cell.userNameLabel.text = repo.user.userName
        
        cell.userThumbImageView.kf.setImage(with: URL(string: repo.user.imgURL))
        cell.userThumbImageView.isUserInteractionEnabled = true
        cell.userThumbImageView.tag = indexPath.row
        
        let tapped:UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(tappedOnImage(_:)))
        tapped.numberOfTapsRequired = 1
        cell.userThumbImageView.addGestureRecognizer(tapped)
        cell.userThumbImageView.tag = indexPath.row
        
        cell.forkNoLabel.text = "\(repo.forksNo) forks"
        cell.issueNoLabel.text = "\(repo.issuesNo) issues"
        cell.watcherNoLabel.text = "\(repo.watchersNo) watchers"
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UIScreen.main.bounds.height * 0.1
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.indexPathSelected = indexPath
        self.presenter.gotoRepoDetailScreen()
    }
    
    func tappedOnImage(_ sender:UITapGestureRecognizer){
        let view = sender.view
        let contentView = view?.superview
        let cell = contentView?.superview
        self.indexPathSelected = tableView.indexPath(for: cell as! UITableViewCell)!
        self.presenter.gotoUserDetailScreen()
    }
    
    @IBAction func segmentedControlValueChanged(_ sender: UISegmentedControl) {
        switch sender.selectedSegmentIndex {
        case 0:  //stars selected
            self.selectedSortMethod = "stars"
            repos.sort{$0.stargazersNo > $1.stargazersNo}
            break
        case 1:   //forks selected
            self.selectedSortMethod = "forks"
            repos.sort{$0.forksNo > $1.forksNo}
            break
        default:  //date selected
            self.selectedSortMethod = "updated"
            repos.sort{$0.datePushed! > $1.datePushed!}
        }
        
        DispatchQueue.main.async {
            self.tableView.reloadData()
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        self.presenter.passDataToNextScene(segue)
    }
}

extension RepoSearchViewController:RepoSearchViewControllerInput{
    
    func displayFetchedRepos(_ repos: [Repo]) {
        self.repos = repos
        
        DispatchQueue.main.async {
            self.tableView.reloadData()
        }
    }
    
    func showWaitingView() {
        let alert = UIAlertController(title: nil, message: "Please wait...", preferredStyle: .alert)
        
        alert.view.tintColor = UIColor.black
        let loadingIndicator = UIActivityIndicatorView(frame: CGRect(x: 10, y: 5, width: 50, height: 50))
        
        loadingIndicator.hidesWhenStopped = true
        loadingIndicator.activityIndicatorViewStyle = .gray
        loadingIndicator.startAnimating()
        
        alert.view.addSubview(loadingIndicator)
        self.navigationController?.present(alert, animated: true, completion: nil)
    }
    
    func hideWaitingView() {
        self.navigationController?.dismiss(animated: true, completion: nil)
    }
    
    func displayErrorView(_ errorMessage: String) {
        let refreshAlert = UIAlertController(title: "Error", message: errorMessage, preferredStyle: .alert)
        
        refreshAlert.addAction(UIAlertAction(title: "OK", style: .default, handler: {(action) in
            refreshAlert.dismiss(animated:true, completion : nil)
        }))
        
        present(refreshAlert, animated: true, completion: nil)
    }
}

extension RepoSearchViewController : UISearchBarDelegate{
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        
        self.presenter.fetchRepos(query:searchController.searchBar.text!, sortMethod: self.selectedSortMethod)
    }
    
}
